File: ios/BockMedia/Domain/UnifiedSearchCoordinator.swift
import Foundation

/// Spotify-style search: suggest @ 1 char, fast grouped search @ 2+.
enum UnifiedSearchCoordinator {
    static let debounceMs: UInt64 = 200_000_000
    static let timeoutSec: TimeInterval = 15
    static let searchLimit = 30
    static let preview = 5

    struct Outcome {
        let response: SearchResponse
        let fromCache: Bool
        let prefixExtension: SearchResponse?
    }

    static func resolveCached(_ query: String) -> Outcome? {
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }
        if let hit = SearchQueryCache.get(trimmed) {
            return Outcome(response: hit, fromCache: true, prefixExtension: nil)
        }
        if trimmed.count >= 2, let prefix = SearchQueryCache.prefixExtension(trimmed) {
            return Outcome(response: prefix, fromCache: false, prefixExtension: prefix)
        }
        return nil
    }

    static func fetch(repository: BockMediaRepository, query: String, source: String?) async throws -> SearchResponse {
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        return try await withThrowingTaskGroup(of: SearchResponse.self) { group in
            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(Self.timeoutSec * 1_000_000_000))
                throw URLError(.timedOut)
            }
            group.addTask {
                if trimmed.count == 1 {
                    return try await repository.searchSuggest(q: trimmed)
                }
                return try await repository.search(
                    q: trimmed,
                    limit: searchLimit,
                    preview: preview,
                    source: source?.isEmpty == false ? source : nil,
                    fast: true,
                    includeResonance: false,
                    includeRooms: false
                )
            }
            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }

    static func userErrorMessage(_ error: Error, remoteOk: Bool) -> String {
        if !remoteOk { return "Search unavailable — check your connection" }
        if (error as? URLError)?.code == .timedOut { return "Search timed out — try again" }
        return error.localizedDescription.isEmpty ? "Search failed" : error.localizedDescription
    }
}

File: android/app/src/main/kotlin/com/bockmedia/console/domain/model/UnifiedSearchCoordinator.kt
package com.bockmedia.console.domain.model

import com.bockmedia.console.data.api.dto.SearchResponse
import com.bockmedia.console.data.repository.BockMediaRepository
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.withTimeout

/** Spotify-style search pipeline: suggest @ 1 char, fast grouped search @ 2+. */
object UnifiedSearchCoordinator {
    const val DEBOUNCE_MS = 200L
    const val TIMEOUT_MS = 15_000L
    const val SEARCH_LIMIT = 30
    const val PREVIEW = 5

    data class Outcome(
        val response: SearchResponse,
        val fromCache: Boolean = false,
        val prefixExtension: SearchResponse? = null,
    )

    /** Resolve cache layers before network. Returns null if a fetch is required. */
    fun resolveCached(query: String): Outcome? {
        val trimmed = query.trim()
        if (trimmed.isEmpty()) return null
        SearchQueryCache.get(trimmed)?.let { return Outcome(it, fromCache = true) }
        if (trimmed.length >= 2) {
            SearchQueryCache.getPrefixExtension(trimmed)?.let { prefix ->
                return Outcome(prefix, prefixExtension = prefix)
            }
        }
        return null
    }

    suspend fun fetch(
        repository: BockMediaRepository,
        query: String,
        source: String? = null,
    ): SearchResponse = withTimeout(TIMEOUT_MS) {
        val trimmed = query.trim()
        when {
            trimmed.length == 1 -> repository.searchSuggest(trimmed)
            else -> repository.search(
                q = trimmed,
                limit = SEARCH_LIMIT,
                preview = PREVIEW,
                source = source?.takeIf { it.isNotBlank() },
                fast = true,
                includeResonance = false,
                includeRooms = false,
            )
        }
    }

    fun cachePut(query: String, response: SearchResponse) {
        SearchQueryCache.put(query.trim(), response)
    }

    fun userErrorMessage(err: Throwable, remoteOk: Boolean): String = when {
        !remoteOk -> "Search unavailable — check your connection"
        err is TimeoutCancellationException -> "Search timed out — try again"
        else -> err.message?.takeIf { it.isNotBlank() } ?: "Search failed"
    }
}
